### ISSUE TITLE
Correct Retry Strategy in Python libraries

### ISSUE THREAD

**Issue by frankyn** [googleapis/google-cloud-python#9298]
Feature request to track work done to fix retry strategy issues in Python libraries.

Issues that have a workaround will be deduped against this bug as @crwilcox, @tritone and I organize the effort.

**Comment by micimize** [comment]
This might be possible with the gcs client currently via
```python
session = requests.Session()
session.mount("https://storage.googleapis.com", requests.adapters.HTTPAdapter(max_retries=5))
client = Client(project=env.GOOGLE_PROJECT, _http=session)
```
Haven't tested though

**Comment by dylancaponi** [comment]
@micimize This is giving me errors (below) on deployment.  Any idea why?

I'm assuming Client here is really:
`from google.cloud import storage`
`storage.Client(...)` ?

Full code:
```
from google.cloud import storage
import requests
session = requests.Session()
session.mount("https://storage.googleapis.com", requests.adapters.HTTPAdapter(max_retries=5))
client = storage.Client(project=PROJECT_ID, _http=session)
```

Error when attempting `blob.download_as_string()`:
```
ERROR: (gcloud.functions.deploy) OperationError: code=3, message=Function failed on loading user code. Error message: 401 GET https://storage.googleapis.com/download/storage/v1/b/<redacted file name>/o/secrets?alt=media: ('Request failed with status code', 401, 'Expected one of', <HTTPStatus.OK: 200>, <HTTPStatus.PARTIAL_CONTENT: 206>)
```

**Comment by micimize** [comment]
@dylancaponi does it work in the same environment without the `_http` argument? 401 means unauthorized, so you probably need to "bind the credentials" like the [docs say the default does](https://googleapis.dev/python/storage/latest/_modules/google/cloud/storage/client.html#Client):
> If not passed, an `_http` object is created that is bound to the `credentials` for the current object

**Comment by micimize** [comment]
They broke the monorepo up, but I found where that happens for reference:
https://github.com/googleapis/python-cloud-core/blob/58c88db4e5c175d52ab796c519dd4a4ed2ec106e/google/cloud/client.py#L157-L160

**Comment by RunOrVeith** [comment]
Since these retry-able errors happen quite frequently, I'd really like for this to be a feature sooner than later. 
Thanks for your efforts!

**Comment by dylancaponi** [comment]
@micimize Yes, the code runs when we don't pass `_http`.  That is how our normal (non-retrying) connection is established.

`builtins.storage_client = storage.Client(project='<redacted>')`

However, that means it's not passing the session that defines 5 retries which I really want to do.  Do you know what the proper approach to that is?

**Comment by micimize** [comment]
@dylancaponi I don't, but what I would do is keep digging through the source until you find `google.auth.transport.requests.AuthorizedSession`. You basically want to extract the auth logic they're using and figure out how to combine it with the `requests.adapters.HTTPAdapter(max_retries=5)` solution, and then pass the resulting session in as `_http`

**Comment by dylancaponi** [comment]
Thanks!  I think I want to do something like `storage_client._http.adapters['https://'].max_retries = 5`.
Will try and report back.

**Comment by dylancaponi** [comment]
We implemented this in production and have seen *zero* errors after 5 days.

Previously we were seeing ~10 errors a day.

This solved our issues.  Thanks @micimize & @pietrodn

```
def add_client_retries(client):
    """ Retry connection to prevent connectionResetError
    """
    adapter = requests.adapters.HTTPAdapter(max_retries=5)
    client._http.mount("https://", adapter)
    client._http._auth_request.session.mount("https://", adapter)
    return client
```
in main.py:
```
from google.cloud import storage
from google.cloud import bigquery
storage_client = add_client_retries(storage.Client())
bq_client = add_client_retries(bigquery.Client())
```
Also, remember to only initialize these once outside of your main() function and reuse them in main().

**Comment by dylancaponi** [comment]
The solution above worked for a month or two but we are now experiencing this issue again ...
How is everyone else faring?
